### ISSUE TITLE
Duplicated selectors are not found when they come from an import

### ISSUE THREAD

**Issue by benoitchantre** [wprig/wprig#407]
## Issue Overview

Duplicate selectors and other errors are not detected because style lint is done before imports.

**Comment by benoitchantre** [comment]
Cc. @ataylorme for feedback.

**Comment by ataylorme** [comment]
I can't reproduce. [`assets/css/src/content.css` in the current commit on `v2.0`](https://github.com/wprig/wprig/blob/425552c5058351e047e4bc0a138c7705bd38bb1f/assets/css/src/content.css) has a duplicate `.alignwide` declaration and I am seeing it reported.



```sh
npm run gulp -- styles                                                                                                               04/2/19 at 2:00:06

> wprig@2.0.0 gulp /Users/andrewtaylor/Sites/ataylorme-wordpress/web/wp-content/themes/wprig
> gulp "styles"

[14:00:26] Requiring external module @babel/register
[14:00:29] Using gulpfile ~/Sites/ataylorme-wordpress/web/wp-content/themes/wprig/gulpfile.babel.js
[14:00:29] Starting 'styles'...

assets/css/src/content.css
264:1	⚠  Unexpected duplicate selector ".alignwide", first used at line 176 (no-duplicate-selectors) [stylelint]
268:1	⚠  Unexpected duplicate selector ".alignfull", first used at line 180 (no-duplicate-selectors) [stylelint]

[14:00:31] Finished 'styles' after 1.84 s
```

**Comment by ataylorme** [comment]
Also if I move `stylelint()` down after `AtImport` in the `postcssPlugins` array so it runs second a lot of warnings show up due to the value being imported into the CSS contents in the stream.

Example:

```sh
assets/css/src/widgets.css
6:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
16:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
21:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
25:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
38:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
43:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
48:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
52:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
57:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
64:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
72:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
80:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
85:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
90:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
95:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
101:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
106:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
112:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
118:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
124:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
131:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
140:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
145:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
152:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
157:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
162:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
166:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
174:1	⚠  Expected empty line before rule (rule-empty-line-before) [stylelint]
185:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
191:1	⚠  Expected empty line before comment (comment-empty-line-before) [stylelint]
```

**Comment by benoitchantre** [comment]
In `content.css`, block selectors are duplicated because they are also imported from `_blocks.css`. This is not detected.

**Comment by ataylorme** [comment]
> Duplicated selectors are not found when they come from an import

Ah, would happen because files prefixed with an underscore, such as `_blocks.css`, are ignored in the CSS source for gulp

```js
src: [
  `${assetsDir}/css/src/**/*.css`,
  // Ignore partial files.
  `!${assetsDir}/css/src/**/_*.css`,
```

And, as you say, `stylelint` happens before the CSS import is processed so the contents of the files aren't in the stream yet when `stylelint` runs

**Comment by ataylorme** [comment]
> And, as you say, stylelint happens before the CSS import is processed so the contents of the files aren't in the stream yet when stylelint runs

Actually, this just effects stylelint not running in `_` prefixed files.

> Duplicated selectors are not found when they come from an import

This is still an issue. [As I said in my comment above](https://github.com/wprig/wprig/issues/407#issuecomment-479201818) running the import first introduces stylelint warnings.

I'm not sure how to resolve that.

**Comment by ataylorme** [comment]
#408 addresses some `stylelint` issues in partial files (prefixed with a `_` and ignored in `gulp.src`) by running `stylelint` as a plugin in `postcss-import`.

CSS files having duplicate selectors from imported files, however, remains unreported.

I've tried quite a few things without success. Maybe @jonathantneal would be kind enough to help us out.

**Comment by mor10** [comment]
I think this is an unavoidable issue as long as we use partials. We have to weigh the pros and cons.

**Comment by ataylorme** [comment]
> I think this is an unavoidable issue as long as we use partials.

I agree. I have been testing #408, which runs `stylelint` after processing imports. 

`_blocks.css` is imported into `content.css`. When I add a selector to `_blocks.css` that already exists in `content.css`, such as `.entry`, is not reported.

Viewing the built CSS file confirms the duplicate entry exists after the import is processed.

Adding a duplicate `.entry` selector directly to `content.css` is reported.

I'm not sure if this is a bug with `stylelint` or not.

**Comment by ataylorme** [comment]
Closed by #408
